### PR TITLE
MONGOCRYPT-476 implement MinCover for int32 and int64

### PR DESCRIPTION
**Pull request by kevinAlbs**
# Summary

This is a translation of the MinCover algorithm from https://github.com/10gen/mongo/pull/7031. It will be used to generate the query payload for Queryable Encryption Range Index.

- Implement MinCover algorithm for `int32_t` and `int64_t`.

# Background & Motivation

The .template.c file is an attempt to reduce code duplication between `uint32_t` and `uint64_t`. If there is a concern it harms readability, it can be split into two files.

Additional `BSON_ASSERT_PARAM` checks were motivated by a recent request to handle NULL arguments consistently (MONGOCRYPT-469).

### CODE REVIEW

File: test/test-mc-range-mincover.c [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+            test->range_min,
+            test->range_max);
+   if (test->min.set) {
+      fprintf (stderr, " min=%" PRId32, test->min.value);
+   }
+   if (test->max.set) {
+      fprintf (stderr, " max=%" PRId32, test->max.value);
+   }
+   fprintf (stderr, " sparsity=%zu", test->sparsity);
+   fprintf (stderr, " expected mincover:\n%s", test->expectMincoverString);
```
Comment (eramongodb) [inline test/test-mc-range-mincover.c:55]: Expected `mincover` does not have begin/end guards, whereas actual `mincover` has begin/end guards. Suggest making these consistent.

**Review by eramongodb — CHANGES_REQUESTED** [review]

File: src/mc-range-mincover-generator.template.c [hunk 2]
```diff
@@ -0,0 +1,167 @@
+// mc-range-mincover-generator.template.c is meant to be included in another
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply src/mc-range-mincover-generator.template.c:1]: Good points. Will add copyright notice and rename to `.h`.

File: src/mc-range-mincover-generator.template.c [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+// SERVER-68600.
+typedef struct {
+   UINT_T _rangeMin;
+   UINT_T _rangeMax;
+   size_t _sparsity;
+   // _maxlen is the maximum bit length of edges in the mincover.
+   size_t _maxlen;
+} T (MinCoverGenerator);
+
+static T (MinCoverGenerator) *
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply src/mc-range-mincover-generator.template.c]: Done. Is using `inline` preferred for short static functions?

File: src/mc-range-mincover-generator.template.c [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+                  ") must be less than or equal to max (%" FMT_UINT_T
+                  ") for range search",
+                  rangeMax,
+                  max);
+      return NULL;
+   }
+   T (MinCoverGenerator) *mcg = bson_malloc0 (sizeof (T (MinCoverGenerator)));
+   mcg->_rangeMin = rangeMin;
+   mcg->_rangeMax = rangeMax;
+   mcg->_maxlen = BITS - T (mc_count_leading_zeros) (max);
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply src/mc-range-mincover-generator.template.c]: Done.

File: src/mc-range-mincover-generator.template.c [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+   mcg->_sparsity = sparsity;
+   return mcg;
+}
+
+static void
+T (MinCoverGenerator_destroy) (T (MinCoverGenerator) * mcg)
+{
+   if (!mcg) {
+      return;
+   }
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply src/mc-range-mincover-generator.template.c]: This is a pattern copied from other `*_destroy()` functions. It is intended to enable freeing NULL on error:

```c
foo_t *f = NULL;

if (!something_that_may_error()) {
    goto fail;
}

f = foo_new ();

fail:
    foo_destroy (f);
```

File: src/mc-range-mincover-generator.template.c [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+   const UINT_T ones = ~UINT_C (0);
+
+   BSON_ASSERT (maskedBits <= BITS);
+   BSON_ASSERT (maskedBits >= 0);
+
+   if (maskedBits == 0) {
+      return value;
+   }
+
+   const UINT_T shift = (BITS - (UINT_T) (maskedBits));
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply src/mc-range-mincover-generator.template.c]: Done.

File: src/mc-range-mincover-generator.template.c [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+}
+
+// applyMask applies a mask of 1 bits starting from the right.
+// Bits 0 to bit-1 are replaced with 1. Other bits are left as-is.
+static UINT_T
+T (applyMask) (UINT_T value, size_t maskedBits)
+{
+   const UINT_T ones = ~UINT_C (0);
+
+   BSON_ASSERT (maskedBits <= BITS);
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply src/mc-range-mincover-generator.template.c]: Done.

File: src/mc-range-mincover-generator.template.c [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+}
+
+char *
+T (MinCoverGenerator_toString) (T (MinCoverGenerator) * mcg,
+                                UINT_T start,
+                                size_t maskedBits)
+{
+   BSON_ASSERT_PARAM (mcg);
+   BSON_ASSERT (maskedBits <= mcg->_maxlen);
+   BSON_ASSERT (maskedBits <= BITS);
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply src/mc-range-mincover-generator.template.c]: Done.

File: src/mc-range-mincover-generator.template.c [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+   BSON_ASSERT (maskedBits >= 0);
+
+   if (maskedBits == mcg->_maxlen) {
+      return bson_strdup ("root");
+   }
+
+   UINT_T shifted = start >> (UINT_T) maskedBits;
+   char *valueBin = T (mc_convert_to_bitstring) (shifted);
+   char *ret = bson_strndup (valueBin + (BITS - mcg->_maxlen + maskedBits),
+                             mcg->_maxlen + maskedBits);
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply src/mc-range-mincover-generator.template.c]: Done.

File: src/mc-range-mincover-generator.template.c [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+   if (blockStart >= mcg->_rangeMin && blockEnd <= mcg->_rangeMax &&
+       T (MinCoverGenerator_isLevelStored) (mcg, maskedBits)) {
+      char *edge = T (MinCoverGenerator_toString) (mcg, blockStart, maskedBits);
+      _mc_array_append_val (c, edge);
+      return;
+   }
+
+   BSON_ASSERT (maskedBits > 0);
+
+   const size_t newBits = maskedBits - 1;
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply src/mc-range-mincover-generator.template.c]: Done.

File: src/mc-range-mincover-generator.template.c [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+   T (MinCoverGenerator_minCoverRec) (mcg, c, blockStart, newBits);
+   T (MinCoverGenerator_minCoverRec)
+   (mcg, c, blockStart | (UINT_T) 1 << (UINT_T) newBits, newBits);
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply src/mc-range-mincover-generator.template.c]: Done.

File: src/mc-range-mincover-generator.template.c [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+                                UINT_T max,
+                                size_t sparsity)
+{
+   BSON_ASSERT_PARAM (mcg);
+   mc_mincover_t *mc = mc_mincover_new ();
+   T (MinCoverGenerator_minCoverRec) (mcg, &mc->mincover, 0, mcg->_maxlen);
+   return mc;
+}
+
+#undef UINT_T
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply src/mc-range-mincover-generator.template.c]: Done.

File: src/mc-range-mincover-generator.template.c [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+   T (MinCoverGenerator_minCoverRec)
+   (mcg, c, blockStart | (UINT_T) 1 << (UINT_T) newBits, newBits);
+}
+
+static mc_mincover_t *
+T (MinCoverGenerator_minCover) (T (MinCoverGenerator) * mcg,
+                                UINT_T rangeMin,
+                                UINT_T rangeMax,
+                                UINT_T max,
+                                size_t sparsity)
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply src/mc-range-mincover-generator.template.c]: Good catch. They are unnecessary. Removed.

File: test/test-mc-range-mincover.c (hunk 1 above)
Comment (kevinAlbs) [reply test/test-mc-range-mincover.c]: Done.

File: src/mc-range-mincover-generator.template.h [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+   BSON_ASSERT_PARAM (mcg);
+   BSON_ASSERT (maskedBits <= mcg->_maxlen);
+   BSON_ASSERT (maskedBits <= (size_t) BITS);
+   BSON_ASSERT (maskedBits >= 0);
+
+   if (maskedBits == mcg->_maxlen) {
+      return bson_strdup ("root");
+   }
+
+   UINT_T shifted = start >> (UINT_T) maskedBits;
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply src/mc-range-mincover-generator.template.h]: Interesting. I originally expected `start >> maskedBits` to have type `size_t`.

Based on https://port70.net/~nsz/c/c99/n1256.html#6.5.7

> The integer promotions are performed on each of the operands. The type of the result is that of the promoted left operand.

IIUC the integer promotions do not promote both operands with respect to one another. I.e. integer promotion occurs on each operand, but both operands do not need to be the same type. Does that match your understanding?

File: test/test-mc-range-mincover.c [hunk 15]
```diff
@@ -0,0 +1,383 @@
...
+            test->range_min,
+            test->range_max);
+   if (test->min.set) {
+      fprintf (stderr, " min=%" PRId64, test->min.value);
+   }
+   if (test->max.set) {
+      fprintf (stderr, " max=%" PRId64, test->max.value);
+   }
+   fprintf (stderr, " sparsity=%zu", test->sparsity);
+   fprintf (stderr, " expected mincover:\n%s", test->expectMincoverString);
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply test/test-mc-range-mincover.c]: Good catch. Updated.

File: src/mc-range-mincover-generator.template.h [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+   const UINT_T ones = ~UINT_C (0);
+
+   BSON_ASSERT (maskedBits <= (size_t) BITS);
+   BSON_ASSERT (maskedBits >= 0);
+
+   if (maskedBits == 0) {
+      return value;
+   }
+
+   const size_t shift = ((size_t) BITS - (UINT_T) (maskedBits));
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply src/mc-range-mincover-generator.template.h]: Done.

**Review by eramongodb — APPROVED** [review]

**Review by vector-of-bool — APPROVED** [review]
Sorry for the delay. Everything LGTM following Ezra's suggestions.

**Pull request merged**